Fixing issue #2009

jurgenvinju opened the pull request:
The SymbolFactory  is/was still implementating the escaping and character encoding of SDF2 and not that of Rascal. This produces weird errors with unicode characters and escapes (\a, \u and \U).

jurgenvinju commented:
Issue #2009 is about character classes like `[A-Z]` but literals seem to need the same fixing:
https://github.com/usethesource/rascal/blob/9b9285ae89da0862168fe96785904ac4de9a1300/src/org/rascalmpl/values/parsetrees/SymbolFactory.java#L203

So let's add some more tests for that and fix them, and then merge the PR when that's done. I'm shifting back to _draft_.

jurgenvinju commented:
This code was still implementating the escaping and character encoding of SDF2 and not that of Rascal.

DavyLandman commented on src/org/rascalmpl/values/parsetrees/SymbolFactory.java:
```diff
@@ -337,31 +337,30 @@ else if (range.isFromTo()) {
-			case '<' : return factory.integer('<');
-			case '>' : return factory.integer('>');
-			case '\\' : return factory.integer('\\');
+				case 't': return factory.integer('\t');
```
do you know if the `'t'` is upgraded to an int, or if the 16bit highest bit of `cha` are dropped (aka downcast to a char)? If the latter, this can cause issues.

jurgenvinju replied:
I'm guessing the former because tihis code has been there for a while and it occurs often in whitespace classes like `[\ \t\n\r]`.

jurgenvinju replied:
https://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html#jls-5.1.2

the char is widened to an int.

DavyLandman replied:
I asked this becauase `cha` went from `char` to `int`. And there could be some high codepoints where only the high 16bits are different from `n`. Like:  [`0x2006E` aka 𠃦 (of which the lower bits are `0x006E` aka `n`)](https://www.compart.com/en/unicode/U+200E6)

jurgenvinju replied:
`int cha` is now the `int String::codepointAt` a given index. This method returns an `int`. And the literal java escapes we used are all below 7-bit ASCII. Their character values are safely widened to 32-bits by adding zeroes. Finally the Unicode escapes are meant to index into 24-bit Unicode codepoints directly (no surrogate encoding).

However, if you want to be smart you can take two low Unicode escapes and construct a high-surrogate/low surrogate pair. This will not work correctly. Probably you end up with two question marks in the console.

jurgenvinju replied:
The code we had before was broken for all the Unicode characters that have a high surrogate in char16 encoding. So this fix will repair that, but if you depended on the buggy behavior you now have a problem. That would be pretty unlikely since the high surrogates were chosen well in char16 encoding. They don't accidentally overlap with commonly used characters.

DavyLandman reviewed (approved):
Looks good, great that you also added new tests 👍🏼

jurgenvinju commented on src/org/rascalmpl/values/parsetrees/SymbolFactory.java:
```diff
@@ -337,31 +337,30 @@ else if (range.isFromTo()) {
-				throw new NotYetImplemented("unicode escape sequence in character class types");
-			}
-			char cha = s.charAt(1);
+		if (s.matches("\\\\[auU][0-9A-F]+")) {
```
I don't think it would be noticeable unless you generate a huge file with lots and lots of character classes.

jurgenvinju commented:
@DavyLandman I added more initially failing tests for the literals in types, such as `{X "\a0A"}*` and the unicode characters in literals and unicode escapes. Fixed easily by reusing the string notation of vallang (which is exactly equal by design). This really ready now. No hurry though.

DavyLandman commented on src/org/rascalmpl/library/lang/rascal/tests/concrete/Character.rsc:
```diff
@@ -49,6 +49,47 @@ test bool charClassOrderedRanges() = (#[a-z A-Z]).symbol == \char-class([range(6
+test bool literalAsciiEscape1() = lit("\n") == #"\a0A".symbol;
+test bool literalAsciiEscape2() = lit("w") == #"\a77".symbol;
+test bool literalUtf16Escape() = lit("\n") == #"\u000A".symbol;
+test bool literalUtf24Escape1() = lit("\n") == #"\U00000A".symbol;
```
nit: it's not "UTF16" and "UTF24". if anything it's utf16 and utf32. or just 16bit & 24bit codepoint

DavyLandman commented on src/org/rascalmpl/values/parsetrees/SymbolFactory.java:
```diff
@@ -197,63 +201,16 @@ private static IValue symbolAST2SymbolConstructor(Sym symbol, boolean lex, Strin
+		try {
+			String lit = ((StringConstant.Lexical) sep).getString();
+			// this should be the exact notation for string literals in vallang
+			IValue string = new StandardTextReader().read(factory, new StringReader(lit));
```
comparing the code, `\\f` is not in vallang `StandardTextReader`. the rest is there it seems.

jurgenvinju commented:
> @jurgenvinju I've fixed 2 bugs I found, and made an extra test case. But I think we have to bring back the escape code. as letting vallang take care breaks support for `\f`. Maybe we should add `\f` support in vallang?

We have to add it to vallang then. In another PR. Let's merge this one because it fixes some serious issues?

jurgenvinju commented:
I merged this into #2002 because of overlap.
